### PR TITLE
Credit card payments refactoring

### PR DESCRIPTION
**Pull request by mingazhev**
1. Divided payfort service to separate services. Signature calculation to SignatureCalculationService,
response parsing to ResponseParser etc.
2. Added new CreditCardPaymentProcessingService to work with credit card payments.
This service does not know about what service is paid and communicates with booking service via interface, passed as method parameter.
Booking service has to give payment amount and process after-payment changes on its own (This is reflected in interface "IServicePaymentsService").

These two steps are in different commits (each commit contains significally changed files)

### CODE REVIEW

File: Api/Services/Accommodations/Bookings/BookingPaymentService.cs [hunk 1]
```diff
@@ -348,13 +324,53 @@ private Task ChangeBookingPaymentStatusToCaptured(Booking booking)
...
             BookingStatusCodes.Pending, BookingStatusCodes.Confirmed
         };
+        
+        public async Task<Result<MoneyAmount>> GetServicePrice(string referenceCode)
+        {
+            var booking = await _context.Bookings.SingleOrDefaultAsync(b => b.ReferenceCode == referenceCode);
+            if(booking == default)
+                return Result.Fail<MoneyAmount>("Could not find booking");
+
+            var agreement = JsonConvert.DeserializeObject<BookingAvailabilityInfo>(booking.ServiceDetails).Agreement;
```
Comment (hasmikharutyunyan) [inline Api/Services/Accommodations/Bookings/BookingPaymentService.cs:334]: can we have here NullReferenceException?

File: Api/Services/Payments/IServicePaymentsService.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+using System.Threading.Tasks;
+using CSharpFunctionalExtensions;
+using HappyTravel.Edo.Api.Models.Payments;
+using HappyTravel.Edo.Data.Payments;
+
+namespace HappyTravel.Edo.Api.Services.Payments.CreditCards
+{
+    public interface IServicePaymentsService
```
Comment (hasmikharutyunyan) [inline Api/Services/Payments/IServicePaymentsService.cs:8]: Why not IPaymentsService?

File: Api/Services/Payments/CreditCards/CreditCardsManagementManagementService.cs [hunk 3]
```diff
@@ -16,9 +16,9 @@
 
 namespace HappyTravel.Edo.Api.Services.Payments.CreditCards
 {
-    public class CreditCardService : ICreditCardService
+    public class CreditCardsManagementManagementService : ICreditCardManagementService
```
Comment (hasmikharutyunyan) [inline Api/Services/Payments/CreditCards/CreditCardsManagementManagementService.cs:19]: CreditCardsManagementService, typo here

File: Api/Services/Payments/Payfort/PayfortResponseParser.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+        private static Result<decimal> FromPayfortAmount(string amountString, string currencyString)
+        {
+            if (!Enum.TryParse<Currencies>(currencyString, out var currency))
+                return Result.Fail<decimal>($"Invalid currency in response: {currencyString}");
+
+            if (!decimal.TryParse(amountString, out var amount))
+                return Result.Fail<decimal>("");
+
+            var result = amount / PayfortConstants.ExponentMultipliers[currency];
```
Comment (hasmikharutyunyan) [inline Api/Services/Payments/Payfort/PayfortResponseParser.cs:77]: Can here be KeyNotFoundException?

File: Api/Services/Payments/Payfort/PayfortResponseParser.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        public Result<T> ParseResponse<T>(JObject response)
+        {
+            var model = response.ToObject<T>(PayfortSerializationSettings.Serializer);
+            return model == null
+                ? Result.Fail<T>($"Invalid payfort payment response: '{response}'")
+                : Result.Ok(model);
+        }
+
+
+        private static Result<decimal> FromPayfortAmount(string amountString, string currencyString)
```
Comment (hasmikharutyunyan) [inline Api/Services/Payments/Payfort/PayfortResponseParser.cs:69]: maybe GetFromPayfortAmount?

File: Api/Services/Accommodations/Bookings/BookingPaymentService.cs (hunk 1 above)
Comment (mingazhev) [reply Api/Services/Accommodations/Bookings/BookingPaymentService.cs:334]: No, I don't think so

File: Api/Services/Payments/CreditCards/CreditCardsManagementManagementService.cs (hunk 3 above)
Comment (mingazhev) [reply Api/Services/Payments/CreditCards/CreditCardsManagementManagementService.cs]: Fixed

File: Api/Services/Payments/Payfort/PayfortResponseParser.cs (hunk 4 above)
Comment (mingazhev) [reply Api/Services/Payments/Payfort/PayfortResponseParser.cs:77]: Looks like it's not very possible

File: Api/Services/Payments/Payfort/PayfortResponseParser.cs (hunk 5 above)
Comment (mingazhev) [reply Api/Services/Payments/Payfort/PayfortResponseParser.cs]: Fixed

File: Api/Services/Payments/Payfort/IPayfortResponseParser.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+using HappyTravel.Edo.Api.Models.Payments.Payfort;
+using Newtonsoft.Json.Linq;
+
+namespace HappyTravel.Edo.Api.Services.Payments.Payfort
+{
+    public interface IPayfortResponseParser
+    {
+        Result<CreditCardPaymentResult> ParsePaymentResponse(JObject response);
+
+        Result<T> ParseResponse<T>(JObject response);
```
Comment (kirillta) [inline Api/Services/Payments/Payfort/IPayfortResponseParser.cs:11]: ResponseParser.ParseResponse again

File: Api/Services/Payments/Payfort/PayfortResponseParser.cs (hunk 4 above)
Comment (kirillta) [reply Api/Services/Payments/Payfort/PayfortResponseParser.cs:77]: I think we must take it into account, because fail on payments is undesirable

Comment (kirillta) [reply Api/Services/Payments/Payfort/PayfortResponseParser.cs:77]: I mean all exception cases

File: Api/Services/Payments/CreditCards/CreditCardPaymentService.cs [hunk 7]
```diff
@@ -1,677 +1,369 @@
...
             
-            bool IsSaveCardNeeded(PaymentResponse response)
+            bool IsSaveCardNeeded(CreditCardPaymentResult response)
                 => request.IsSaveCardNeeded && response.Status != CreditCardPaymentStatuses.Failed;
+            
+            Task SaveCard(CreditCardPaymentResult _) => _creditCardsManagementService.Save(request.CardInfo, request.Token, customer);
+            
+            async Task<Result<PaymentResponse>> StorePaymentResults(CreditCardPaymentResult paymentResult)
+            {
+                var card = request.IsSaveCardNeeded
```
Comment (kirillta) [inline Api/Services/Payments/CreditCards/CreditCardPaymentService.cs:77]: What means "needed"?

Comment (mingazhev) [reply Api/Services/Payments/CreditCards/CreditCardPaymentService.cs:77]: I meant that user asked to save the card.
Another option is IsSaveCardRequested, or maybe you have other suggestions?

File: Api/Services/Payments/CreditCards/ICreditCardPaymentProcessingService.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
+using System.Threading.Tasks;
+using CSharpFunctionalExtensions;
+using HappyTravel.Edo.Api.Models.Payments;
+using Newtonsoft.Json.Linq;
+
+namespace HappyTravel.Edo.Api.Services.Payments.CreditCards
+{
+    public interface ICreditCardPaymentProcessingService
+    {
+        Task<Result<PaymentResponse>> AuthorizeMoney(NewCreditCardPaymentRequest request,
```
[the remark this answers is not in the document]
Comment (mingazhev) [reply Api/Services/Payments/CreditCards/ICreditCardPaymentProcessingService.cs:10]: I thought about 'Authorize' as user authorization, which can be confusing. But I think you're right, I'll change

**Pull request merged**